technologydata/sources.py

```diff
@@ -208,6 +220,424 @@ def process(self, trusted_execution: bool = False) -> None:
+        based on the outcome of the download operation.
+
+        Parameters
+        ----------
+        url_archived : str
+            The URL of the archived file to be downloaded.
+
+        save_path : pathlib.Path
+            The path where the downloaded file will be saved, including the file name.
+
+        Returns
+        -------
+        pathlib.Path | None
+            The path where the file has been saved if the download is successful, or None
+            if an error occurs during the download process.
+
+        Raises
+        ------
+        requests.exceptions.RequestException
+            If there is an issue with the HTTP request, an error is logged, and None is returned.
+
+        """
+        try:
+            response = requests.get(url_archived)
+            response.raise_for_status()  # Check for HTTP errors
+
+            with open(save_path, "wb") as file:
+                file.write(response.content)
+
+            logger.info(f"File downloaded successfully and saved to {save_path}")
+            return save_path
+        except requests.exceptions.RequestException as e:
+            logger.error(f"An error occurred during file download: {e}")
+            return None
+
+    @staticmethod
+    def change_datetime_format(
+        input_datetime_string: str,
+        input_datetime_format: str,
+        output_datetime_format: str,
+    ) -> str | None:
+        """
+        Change the format of a given datetime string from one format to another. This method takes a
+        datetime string and its current format, then converts it to a specified output format.
+        If the input string does not match the provided input format, it logs an error and returns None.
+
+        Parameters
+        ----------
+        input_datetime_string : str
+            datetime string that needs to be reformatted
+
+        input_datetime_format : str
+            format of the input datetime string, following the strftime format codes
+
+        output_datetime_format : str
+            desired format for the output datetime string, following the strftime format codes.
+
+        Returns
+        -------
+           str | None
+               reformatted datetime string if successful, otherwise None
+
+        Raises
+        ------
+        ValueError
+            If the input datetime string does not match the input format.
+
+        """
+        try:
+            dt = datetime.strptime(input_datetime_string, input_datetime_format)
+            logger.info(
+                f"The datetime string follows the format {input_datetime_format}"
+            )
+            output_datetime_string = dt.strftime(output_datetime_format)
+            logger.info(f"The format is now changed to {output_datetime_format}")
+            return output_datetime_string
+        except ValueError as e:
+            logger.info(f"Error during datetime formatting: {e}")
```

Comment:
```suggestion
            logger.error(f"Error during datetime formatting: {e}")
```

or follow this one?
https://stackoverflow.com/questions/6062576/adding-information-to-an-exception/46091127#46091127

Reply:
+ add unit test for the exception :)